File: DESTINY_search_space.py
import numpy as np
from itertools import product


def generate_solutions(size_in_mb):
    solutions = []
    # Define the search space
    capacity_divisors = [i for i in range(1, size_in_mb+1) if size_in_mb % i == 0] #capacity in MB or it's divisors
    word_width_options = [2**i for i in range(3, 11)]  # 8 to 1024 wordwidts
    #mux_options_amp = np.arange(1, 257) # only integer values starting from 1 to 256
    #mux_options2 = np.arange(1, 257) # only integer values starting from 1 to 256
    #mux_options3 = np.arange(1, 257) # only integer values starting from 1 to 256
    mux_options_amp = [1,2,4,8,64,128,256] 
    mux_options2 = [1,2,4,8,64,128,256]  
    mux_options3 = [1,2,4,8,64,128,256] 
    # Generate square mat sizes that divide 144 evenly
    mat_sizes = [i for i in range(1, 13) if 144 % (i * i) == 0] #12x12 mats or its equal subcombinations
    subbaray_set = [1,2,4,6,8,10,12]
    
    # Generate all combinations
    for (capacity_in_mb, word_width, MuxSenseAmp, MuxOutputLev1, MuxOutputLev2, mat_size, num_subarray, ) in product(
        capacity_divisors, word_width_options, mux_options_amp, mux_options2, mux_options3, mat_sizes, subbaray_set):
        
        num_row_mat = mat_size
        num_col_mat = mat_size
       
        # Fixed values
        associativity = 1 #RAM so it's 1
        stackdiecount = 1 #3D stacking
        num_row_per_set = 1 #RAM so it's 1
        #num_subarray = 1
        
        # Calculations
        capacity = capacity_in_mb * 2**20 * 8 #convert capacity to bits
        num_way_per_row = int(associativity/num_row_per_set)
        num_mats = num_row_mat * num_col_mat
        num_col_mat_active = num_col_mat
        num_row_mat_active = num_row_mat
        N_active_subarrays = num_subarray
        
        Mux_levels = MuxSenseAmp * MuxOutputLev1 * MuxOutputLev2
        num_address_bits = int(np.log2(capacity/word_width/associativity/stackdiecount))
        num_address_bits_mat = num_address_bits - int(np.log2(num_mats/(num_col_mat_active*num_row_mat_active)))
        num_address_bits_gating = int(np.log2(num_subarray/N_active_subarrays))
        num_address_bits_subarray = num_address_bits_mat - num_address_bits_gating
        h = int(np.log2(num_col_mat_active))
        v = int(np.log2(num_row_mat_active))
        row = ((2**(num_address_bits_subarray) * associativity) / Mux_levels * num_way_per_row)
        col = ((word_width / 2**(h+v)) / N_active_subarrays) * Mux_levels * num_way_per_row
        size = num_mats * num_subarray * row * col * stackdiecount
        
        
        # Constrain: capacity = size and col and row are positive integers
        if abs(capacity - size) < 1e-6 and row.is_integer() and col.is_integer() and int(row) > 0 and int(col) > 0:  #small threshold for floating-point comparison
            solutions.append({
                'capacity_in_mb': capacity_in_mb,
                'word_width': word_width,
                'MuxSenseAmp': MuxSenseAmp,
                'MuxOutputLev1': MuxOutputLev1,
                'MuxOutputLev2': MuxOutputLev2,
                'num_mats_col': num_col_mat,
                'num_mats_row': num_row_mat,
                'size': size,
                'row': row,
                'col': col,
                'num_address_bits_subarray': num_address_bits_subarray,
                'num_subbarays': N_active_subarrays
            })
    
    return solutions

# Generate and print solutions
solutions = generate_solutions(size_in_mb = 144)

print(f"Total number of solutions: {len(solutions)}")

#example solutions: 
print(solutions[int(len(solutions)/2)])

# Filter solutions to get unique values of capacity_in_mb and num_mats_col
unique_capacity_in_mb = set()
unique_num_mats_col = set()
unique_word_width = set()

for solution in solutions:
    unique_capacity_in_mb.add(solution['capacity_in_mb'])
    unique_num_mats_col.add(solution['num_mats_col'])
    unique_word_width.add(solution['word_width'])


print(f"Possible values of capacity_in_mb: {sorted(unique_capacity_in_mb)}")
print(f"Possible mat organisation n x n: {sorted(unique_num_mats_col)}")
print(f"Possible values of word_width: {sorted(unique_word_width)}")

File: DESTINY_search_space.m
function DESTINY_search_space()
    tic; % Start timer
    solutions = generate_solutions(144); %desired capacity in MB
    toc; % End timer

    fprintf('Total number of solutions: %d\n', length(solutions));

    if ~isempty(solutions)
        disp('Example solution:');
        disp(solutions(6900));
    end

    % Filter solutions to get unique values
    unique_capacity_in_mb = unique([solutions.capacity_in_mb]);
    unique_num_mats_col = unique([solutions.num_mats_col]);
    unique_word_width = unique([solutions.word_width]);

    fprintf('Possible values of capacity_in_mb: ');
    disp(sort(unique_capacity_in_mb));

    fprintf('Possible mat organisation n x n: ');
    disp(sort(unique_num_mats_col));

    fprintf('Possible values of word_width: ');
    disp(sort(unique_word_width));
end

function solutions = generate_solutions(size_in_mb)
    % Define the search space
    capacity_divisors = find(rem(size_in_mb, 1:size_in_mb) == 0);
    word_width_options = 2.^(3:10);  % 8 to 1024 wordwidts
    mux_options = [1,2,4,8,64,128,256];
    mat_sizes = find(rem(144, (1:12).^2) == 0);
    subbaray_set = [1,2,4,6,8,10,12];
    
    % Generate all combinations
    [C1,C2,C3,C4,C5,C6,C7] = ndgrid(capacity_divisors, word_width_options, mux_options, ...
        mux_options, mux_options, mat_sizes, subbaray_set);
    
    capacity_in_mb = C1(:);
    word_width = C2(:);
    MuxSenseAmp = C3(:);
    MuxOutputLev1 = C4(:);
    MuxOutputLev2 = C5(:);
    mat_size = C6(:);
    num_subarray = C7(:);
    
    % Fixed values
    associativity = 1;
    stackdiecount = 1;
    num_row_per_set = 1;
    
    % Vectorized calculations
    capacity = capacity_in_mb * 2^20 * 8;
    num_way_per_row = floor(associativity./num_row_per_set);
    num_mats = mat_size.^2;
    num_col_mat_active = mat_size;
    num_row_mat_active = mat_size;
    N_active_subarrays = num_subarray;
    
    Mux_levels = MuxSenseAmp .* MuxOutputLev1 .* MuxOutputLev2;
    num_address_bits = floor(log2(capacity./word_width/associativity/stackdiecount));
    num_address_bits_mat = num_address_bits - floor(log2(num_mats./(num_col_mat_active.*num_row_mat_active)));
    num_address_bits_gating = floor(log2(num_subarray./N_active_subarrays));
    num_address_bits_subarray = num_address_bits_mat - num_address_bits_gating;
    h = floor(log2(num_col_mat_active));
    v = floor(log2(num_row_mat_active));
    row = ((2.^num_address_bits_subarray * associativity) ./ Mux_levels .* num_way_per_row);
    col = ((word_width ./ 2.^(h+v)) ./ N_active_subarrays) .* Mux_levels .* num_way_per_row;
    calculated_size = num_mats .* num_subarray .* row .* col * stackdiecount;
    
    % Apply constraints
    valid = abs(capacity - calculated_size) < 1e-6 & mod(row,1) == 0 & mod(col,1) == 0 & row > 0 & col > 0;
    
    % Create solution struct array
    solutions = struct('capacity_in_mb', num2cell(capacity_in_mb(valid)), ...
                       'word_width', num2cell(word_width(valid)), ...
                       'MuxSenseAmp', num2cell(MuxSenseAmp(valid)), ...
                       'MuxOutputLev1', num2cell(MuxOutputLev1(valid)), ...
                       'MuxOutputLev2', num2cell(MuxOutputLev2(valid)), ...
                       'num_mats_col', num2cell(mat_size(valid)), ...
                       'num_mats_row', num2cell(mat_size(valid)), ...
                       'size', num2cell(calculated_size(valid)), ...
                       'row', num2cell(row(valid)), ...
                       'col', num2cell(col(valid)), ...
                       'num_address_bits_subarray', num2cell(num_address_bits_subarray(valid)), ...
                       'num_subbarays', num2cell(num_subarray(valid)));
end